stdlib/public/Reflection/TypeLowering.cpp

```diff
@@ -1890,33 +2012,76 @@ class EnumTypeInfoBuilder {
+            auto PayloadSpareBitMaskByteOffset = MPEDescriptor->getPayloadSpareBitMaskByteOffset();
+            auto SpareBitMask = MPEDescriptor->getPayloadSpareBits();
+            BitMask spareBitsMask(PayloadSize, SpareBitMask,
+                                  PayloadSpareBitMaskByteCount, PayloadSpareBitMaskByteOffset);
+
+            if (!spareBitsMask.isZero()) {
+#if !defined(NDEBUG) && 0
+              // DEBUGGING: compare the locally-computed spare bit mask to the
+              // one we got from the compiler.  If they're different, then
+              // either the compiler is emitting the wrong thing or the
+              // local runtime computation isn't quite right.
+              BitMask locallyComputedSpareBitsMask(PayloadSize);
+              auto mpePointerSpareBits = TC.getBuilder().getMultiPayloadEnumPointerMask();
+              auto locallyComputedSpareBitsMaskIsValid
+                = populateSpareBitsMask(Cases, locallyComputedSpareBitsMask, mpePointerSpareBits);
+              assert(locallyComputedSpareBitsMaskIsValid);
+              assert(locallyComputedSpareBitsMask == spareBitsMask);
+#endif
+
+              // Use compiler-provided spare bit information
+              return TC.makeTypeInfo<MultiPayloadEnumTypeInfo>(
+                Size, Alignment, Stride, NumExtraInhabitants,
+                BitwiseTakable, Cases, spareBitsMask);
+            } else {
+              // The MPE descriptor doesn't make sense: It has a spare bit mask,
+              // but that mask is empty?  If this ever happens, fall through to
+              // the local calculation.
+            }
+          } else {
             // If there are no spare bits, use the "simple" tag-only implementation.
             return TC.makeTypeInfo<SimpleMultiPayloadEnumTypeInfo>(
               Size, Alignment, Stride, NumExtraInhabitants,
               BitwiseTakable, Cases);
-          } else {
-            // General case using a mix of spare bits and extra tag
-            return TC.makeTypeInfo<MultiPayloadEnumTypeInfo>(
-              Size, Alignment, Stride, NumExtraInhabitants,
-              BitwiseTakable, Cases, spareBitsMask);
           }
         }
-*/
 
-        // Without spare bit mask info, we have to leave this particular
-        // enum as "Unsupported", meaning we will not be able to project
-        // cases or evaluate XIs.
-        return TC.makeTypeInfo<UnsupportedEnumTypeInfo>(
-          Size, Alignment, Stride, NumExtraInhabitants,
-          BitwiseTakable, EnumKind::MultiPayloadEnum, Cases);
+        // If there was no compiler data, try computing the mask ourselves
+        // (This is less robust, but necessary to support images from older
+        // compilers.)
+        BitMask spareBitsMask(PayloadSize);
+        auto mpePointerSpareBits = TC.getBuilder().getMultiPayloadEnumPointerMask();
+        auto validSpareBitsMask = populateSpareBitsMask(Cases, spareBitsMask, mpePointerSpareBits);
```

Comment:
@mikeash @al45tair Here's where I now ask the TypeRefBuilder to fetch the MPE pointer mask from the target and use that to compute the enum spare bit mask.